### PR TITLE
Check entry empty state to ensure GC eligible

### PR DESCRIPTION
**Pull request by rallen090**
**What this PR does / why we need it**:

Before removing in-memory series from the index, we need to be sure that they've been flushed to disk. Before the change in the PR, the proxy for determining this was based on if a given blockStart had been bootstrapped - the logic being that block must have been flushed to disk for it to have been bootstrapped (i.e. `blockTickResult.NumSegmentsBootstrapped != 0`).

The problem w/ the above approach is that it does not capture cold writes, since those could be written to old blockStarts, still be in-memory and not cold flushed yet to disk, but the bootstrapped state would still be true.

Instead, we now check that a given series is truly "empty" (i.e. no TSDB data nor index data for that series in-memory) to be eligible for GC. One challenge with this check is that we only keep track of the state (a) "is TSDB data still in-memory" and not (b) "is index data still in-memory". In other words, we can be sure that (a) is true but cannot be sure (b) is true. So this PR also makes it so that the state we track asserts that both (a) and (b) are true. The way we do this is by marking this state only after both TSDB and index (warm and cold) flushes are truly complete.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/dbnode/storage/flush.go [hunk 1]
```diff
@@ -159,26 +170,48 @@ func (m *flushManager) Flush(startTime xtime.UnixNano) error {
...
-	return multiErr.FinalError()
+	err = multiErr.FinalError()
+
+	// Mark all flush states at the very end to ensure this
+	// happens after both data and index flushing.
+	// TODO: if this matching approach of data + index ns+shard+time is needed, then reimplement this
+	// more efficiently w/ hashing for constant lookup cost.
+	for _, f := range flushes {
+		for _, s := range f.shardFlushes {
+			for _, ff := range indexFlushes {
```
Comment (rallen090) [inline src/dbnode/storage/flush.go:186]: FYI originally I was just using `flushes` and not checking it against `indexFlushes` assuming that anything that could have been "data" flushed must have been "index" flushes since it happens after. That assumption maybe isn't correct, though, and so now this is ensuring that marking a flush requires that the ns+shard+time happened in the warmFlush + indexFlush path.

This is just hacked together atm to test it out quickly and so will need to refactor if we do need this.

Comment (rallen090) [reply src/dbnode/storage/flush.go:186]: Another part I suspect might be incorrect here is these time comparisons - given that index block size and block size usually differ?

Comment (rallen090) [reply src/dbnode/storage/flush.go]: Out of date - since have converted to using maps + using either index or data as the signal for marking.

File: src/dbnode/storage/flush.go [hunk 2]
```diff
@@ -54,6 +54,20 @@ const (
 	flushManagerIndexFlushInProgress
 )
 
+type namespaceFlushes map[string]namespaceFlush
+
+type shardFlushes map[shardFlush]bool
+
+type namespaceFlush struct {
+	namespace    databaseNamespace
+	shardFlushes map[shardFlush]bool
```
Comment (robskillington) [inline src/dbnode/storage/flush.go:63]: I would change this datastructure to:

1. Use only native Go types as the key and avoid using pointer types as part of the key for the map (the comparison of pointer types is not always as you'd expect especially when using interfaces, which we are here).
2. Use a map value to store the pointer types you're trying to access (since don't want to start it part of the key).

```go
// ...
  shardFlushes map[shardFlushKey]shardFlushValue
// ...

type shardFlushKey struct {
  shardID uint32
  blockStart xtime.UnixNano
}

type shardFlushValue struct {
  shard databaseShard
}
```

File: src/dbnode/storage/flush.go [hunk 3]
```diff
@@ -361,18 +430,30 @@ func (m *flushManager) flushNamespaceWithTimes(
...
 			detailedErr := fmt.Errorf("namespace %s failed to flush data: %v",
 				ns.ID().String(), err)
 			multiErr = multiErr.Add(detailedErr)
+		} else {
+			for _, s := range shards {
+				flushes[shardFlush{
+					shard: s,
+					time:  t,
+				}] = true
+			}
```
Comment (robskillington) [inline src/dbnode/storage/flush.go:450]: nit: For more idiomatic Go better to do a continue above so you don't need an else here:
```go
		shards, err := ns.WarmFlush(t, flushPreparer)
		if err != nil {
			detailedErr := fmt.Errorf("namespace %s failed to flush data: %v",
				ns.ID().String(), err)
			multiErr = multiErr.Add(detailedErr)
			continue
		}

		for _, s := range shards {
			flushes[shardFlush{
				shard: s,
				time:  t,
			}] = true
		}
```

As per Effective Go:
https://golang.org/doc/effective_go#if
> In the Go libraries, you'll find that when an if statement doesn't flow into the next statement—that is, the body ends in break, continue, goto, or return—the unnecessary else is omitted.

File: src/dbnode/storage/index.go [hunk 4]
```diff
@@ -1092,10 +1074,17 @@ func (i *nsIndex) WarmFlush(
 				zap.Error(err),
 				zap.Time("blockStart", block.StartTime().ToTime()),
 			)
+		} else {
+			for _, s := range shards {
+				shardFlushes[shardFlush{
+					shard: s,
+					time:  block.StartTime(),
+				}] = true
+			}
```
Comment (robskillington) [inline src/dbnode/storage/index.go:1083]: Same here, would avoid the else with a continue above (as per Effective Go):
```suggestion
			continue
		}
		
		for _, s := range shards {
			shardFlushes[shardFlush{
				shard: s,
				time:  block.StartTime(),
			}] = true
		}
```

File: src/dbnode/storage/namespace.go [hunk 5]
```diff
@@ -1231,12 +1232,18 @@ func (n *dbNamespace) WarmFlush(
 			detailedErr := fmt.Errorf("shard %d failed to flush data: %v",
 				shard.ID(), err)
 			multiErr = multiErr.Add(detailedErr)
+
+			// Errors so mark as failed here.
+			// We mark as success later in time to ensure that happens after index flushing as well.
+			shard.MarkWarmFlushStateSuccessOrError(blockStart, err)
+		} else {
+			flushedShards = append(flushedShards, shard)
 		}
```
Comment (robskillington) [inline src/dbnode/storage/namespace.go:1241]: As per other suggestions on the PR - use an early return here as per Effective Go?
```suggestion
			// Errors so mark as failed here.
			// We mark as success later in time to ensure that happens after index flushing as well.
			shard.MarkWarmFlushStateSuccessOrError(blockStart, err)
			continue
		}
		
		flushedShards = append(flushedShards, shard)
```

File: src/dbnode/storage/series/lookup/entry.go [hunk 6]
```diff
@@ -231,23 +223,9 @@ func (entry *Entry) IfAlreadyIndexedMarkIndexSuccessAndFinalize(
...
-			result.IndexedBlockStartsRemoved++
-			continue
-		}
-		result.IndexedBlockStartsRemaining++
-	}
-	entry.reverseIndex.Unlock()
-	return result
+// IsEmpty returns true if the entry has no in-memory series data.
+func (entry *Entry) IsEmpty() bool {
+	return entry.Series.IsEmpty()
```
Comment (robskillington) [inline src/dbnode/storage/series/lookup/entry.go:228]: Instead of just doing `entry.Series.IsEmpty()` we will need to keep the idea of `RelookupAndIncrementReaderWriterCount` IMO which would re-lookup the latest lookup.Entry in the shard map.

There's a few reasons for this and we saw real problems without relooking up the entry. I can explain it more in depth if required but the problem is that sometimes a lookup.Entry is created for a new series which is passed off to the indexing queue but then a race of two datapoints for a series that doesn't exist yet causes only a single lookup.Entry to exist in the final shard map and the other one becomes orphaned (which could look empty but the lookup.Entry that made it into the shard map will have a series that is not actually empty).

That's why the caller (from within `mutable_segments.go`) should perform perhaps something like the following:
```go
			isEmpty, ok := entry.RelookupAndReturnIsEmpty()
			if !ok {
				// Should not happen since shard will not expire until
				// no more block starts are indexed.
				// We do not GC this series if shard is missing since
				// we open up a race condition where the entry is not
				// in the shard yet and we GC it since we can't find it
				// due to an asynchronous insert.
				return true
			}

			return !isEmpty
```

Comment (robskillington) [reply src/dbnode/storage/series/lookup/entry.go:228]: Also we should probably add a metric for if the edge case happens using the `instrument.EmitAndLogInvariantError(...)` so that this panics scenario tests if this edge case ever starts occurring.

Comment (rallen090) [reply src/dbnode/storage/series/lookup/entry.go:228]: Interesting wasn't aware of this edge case but it makes sense. Will add back this safeguard.

File: src/dbnode/storage/flush.go (hunk 2 above)
Comment (rallen090) [reply src/dbnode/storage/flush.go:63]: Ah good idea - makes sense to key off the primitives. Since we only have one value I think we can use just this key struct for simplicity.

File: src/dbnode/storage/series/lookup/entry.go (hunk 6 above)
Comment (rallen090) [reply src/dbnode/storage/series/lookup/entry.go]: PR updated accordingly

File: src/dbnode/storage/flush.go [hunk 7]
```diff
@@ -159,26 +174,69 @@ func (m *flushManager) Flush(startTime xtime.UnixNano) error {
...
+	// Mark all flushed shards as such.
+	// If index is not enabled, then a shard+blockStart is "flushed" if the data has been flushed.
+	// If index is enabled, then a shard+blockStart is "flushed" if the data AND index has been flushed.
+	for _, n := range namespaces {
+		var (
+			indexEnabled  = n.Options().IndexOptions().Enabled()
+			flushedShards map[shardFlush]bool
+		)
+		if indexEnabled {
+			flushesForNs, ok := indexFlushes[n.ID().String()]
```
[the remark this answers is not in the document]
Comment (rallen090) [reply src/dbnode/storage/flush.go:193]: Data flushes always precede index flushes, so it is safe to assume that the marking of a "full" flush can be indicated by the "index" flush. If the index is disabled, though, then we just use the data flush as the indicator.

[the remark this answers is not in the document]
Comment (rallen090) [reply src/dbnode/storage/flush.go:193]: This is something I am double checking w/ rob though so stay tuned. Assuming what I just said is true, I can add a comment here explaining the logic.

File: src/dbnode/storage/index.go [hunk 8]
```diff
@@ -1074,18 +1073,16 @@ func (i *nsIndex) WarmFlush(
...
 		}
 
-		for _, s := range shards {
-			flushes[shardFlushKey{
-				shardID:    s.ID(),
-				blockStart: block.StartTime(),
-			}] = s
+		for _, t := range i.blockStartsFromIndexBlockStart(block.StartTime()) {
+			for _, s := range shards {
+				s.MarkWarmIndexFlushStateSuccessOrError(t, err)
```
Comment (nbroyles) [inline src/dbnode/storage/index.go:1080]: So, IIUC, `blockStartsFromIndexBlockStart` returns data block starts between index block start and index block start + index block size. This is important because index block size >= data block size. If that's true, why do we need to `MarkWarmIndexFlushStateSuccessOrError` for each data block start time?

File: src/dbnode/storage/shard.go [hunk 9]
```diff
@@ -1937,6 +1951,23 @@ func (s *dbShard) UpdateFlushStates() {
...
+
+	// expose for getting all info files
+	indexFlushedBlockStarts := s.reverseIndex.WarmFlushedBlockStarts()
+	for _, blockStart := range indexFlushedBlockStarts {
+		// Index block size is wider than data block size, so we want to set all data blockStarts
+		// within the range of a given index blockStart
+		for at := blockStart; at < blockStart.Add(indexBlockSize); at = at.Add(blockSize) {
+			currState := s.flushStateNoBootstrapCheck(at)
+			if currState.WarmStatus.DataFlushed != fileOpSuccess {
+				s.markWarmIndexFlushStateSuccess(at)
```
Comment (nbroyles) [inline src/dbnode/storage/shard.go:1966]: You don't do this when `DataFlushed == fileOpSuccess` because that would mean we've already perform `markWarmIndexFlushStateSuccess`, yes?

File: src/dbnode/storage/index.go (hunk 8 above)
Comment (rallen090) [reply src/dbnode/storage/index.go:1080]: That's correct. And the reason is because `MarkWarmIndexFlushStateSuccessOrError` is still marking data blockStarts (not index blockStarts), but each block now just have a flag for both data and index. This is because the state we are tracking is always still at the data block size not index block size. Eg we have 1h block and 2h indexBlock. In-mem we now have:
```
1pm: {dataFlushed: bool, indexFlushed: bool}
2pm: {dataFlushed: bool, indexFlushed: bool}
3pm: {dataFlushed: bool, indexFlushed: bool}
4pm: {dataFlushed: bool, indexFlushed: bool}
...
```
where determining if a given block is eligible for GC we check both `dataFlushed && indexFlushed`. So when an index flush occurs, that actually in this case covers 2 blocks.

This is a somewhat confusing nuance though so if you have any suggestions are making this easier to understand let me know.

File: src/dbnode/storage/shard.go (hunk 9 above)
Comment (rallen090) [reply src/dbnode/storage/shard.go]: Ah this is a bug. Should be checking `IndexFlushed`

File: src/dbnode/storage/index.go (hunk 8 above)
Comment (nbroyles) [reply src/dbnode/storage/index.go:1080]: Yeah, may be worth adding a comment here just clarify
